Stack/Opc.Ua.Core/Security/Certificates/CertificateValidator.cs

```diff
@@ -669,61 +708,53 @@ private static void LogInnerServiceResults(LogLevel logLevel, ServiceResult resu
         /// <summary>
         /// Saves the certificate in the rejected certificate store.
         /// </summary>
-        private void SaveCertificate(X509Certificate2 certificate)
+        private Task SaveCertificateAsync(X509Certificate2 certificate, CancellationToken ct = default)
         {
-            SaveCertificates(new X509Certificate2Collection { certificate });
+            return SaveCertificatesAsync(new X509Certificate2Collection { certificate }, ct);
         }
 
         /// <summary>
         /// Saves the certificate chain in the rejected certificate store.
+        /// Times out after 5 seconds waiting to gracefully reduce high CPU load.
         /// </summary>
-        private void SaveCertificates(X509Certificate2Collection certificateChain)
+        private async Task SaveCertificatesAsync(X509Certificate2Collection certificateChain, CancellationToken ct = default)
         {
+            // max time to wait for semaphore
+            const int kSaveCertificatesTimeout = 5000;
+
+            var rejectedCertificateStore = m_rejectedCertificateStore;
+            if (rejectedCertificateStore == null)
+            {
+                return;
+            }
+
             try
             {
-                m_semaphore.Wait();
+                await m_semaphore.WaitAsync(kSaveCertificatesTimeout, ct).ConfigureAwait(false);
 
-                if (m_rejectedCertificateStore != null)
+                try
                 {
                     Utils.LogTrace("Writing rejected certificate chain to: {0}", m_rejectedCertificateStore);
+
+                    ICertificateStore store = m_rejectedCertificateStore.OpenStore();
                     try
```

Comment:
Other places use the "using()" statement when opening the store, that should get rid of the third imbricated try block